### ISSUE TITLE
[HxInputDate] [HxInputDateRange] Changing value causes two calls of EditContext.OnFieldChanged

### ISSUE THREAD

**Issue by jirikanda** [havit/Havit.Blazor#787]
## Reprosteps
* Handle EditContext.OnFieldChanged event
* Click on a date in the calendar.
* EditContext.OnFieldChanged event handler is called twice

## Why?
(both HxInputDate&lt;TValue&gt; and also HxInputDate**Internal**&lt;TValue&gt; inherit from InputBase&lt;TValue&gt;)

### First call
* HxInputDate**Internal**&lt;TValue&gt;.HandleValueChanged is called.
* HxInputDate**Internal**&lt;TValue&gt;.HandleValueChanged sets the CurrentValueAsString.
* CurrentValueAsString setter sets InputBase&lt;TValuegt;.CurrentValue.
* InputBase&lt;TValue&gt;.CurrentValue calls ValueChanged (important to understand the the second call)
* The InputBase&lt;TValue&gt;.CurrentValue setter calls EditContext.NotifyFieldChanged for the first time.

### Second call
* HxInputDate&lt;Value&gt; handles ValueChanged event
* It sets the CurrentValue property (see BuildRenderInputCore method, sequence number 101).
* The InputBase&lt;TValue&gt;.CurrentValue setter calls (again) EditContext.NotifyFieldChanged.

### Posible solutions (not sure right now)
* only  HxInputDate&lt;TValue&gt; should derive from InputBase&lt;TValue&gt;
* OR HxInputDate&lt;TValue&gt; should not Set CurrentValue, but it should fire ValueChanged event

(I guess this occurs also in HxInputDateRange while it uses the same code structure.)

**Comment by hakenr** [comment]
> only HxInputDate<TValue> should derive from InputBase<TValue>

My preference (same in other similar components).

**Comment by jirikanda** [comment]
@hakenr, I am not familiar with the current code structure, there for I am not able to fix the issue (until someone explains me the thoughts behind). 
ie.
```
	private void HandleValueChanged(string newInputValue)
	{
#if NET8_0_OR_GREATER
		CurrentValueAsString = newInputValue;
#else
		// HandleValueChanged is used instead of TryParseValueFromString
		// When TryParseValueFromString is used (pre net8), invalid input is replaced by previous value.		
		bool parsingFailed;
		_validationMessageStore.Clear(FieldIdentifier);

		if (HxInputDate<DateTime>.TryParseDateTimeOffsetFromString(newInputValue, null, out var date))
		{
			parsingFailed = false;
			CurrentValue = GetValueFromDateTimeOffset(date);
		}
		else
		{
			parsingFailed = true;
			_validationMessageStore.Add(FieldIdentifier, ParsingErrorMessageEffective);
		}

		// We can skip the validation notification if we were previously valid and still are
		if (parsingFailed || _previousParsingAttemptFailed)
		{
			EditContext.NotifyValidationStateChanged();
			_previousParsingAttemptFailed = parsingFailed;
		}
#endif
	}
```

Why are the so different NET8 vs pre-NET8? How to test/debug the differences?

**Comment by hakenr** [comment]
@jirikanda Before .NET 8, the `SetUpdatesAttributeName()` method was missing, and some of our components, such as `HxInputDate`, compensated it by using our own logic. In .NET 8, the new `InputBase` implementation incorporates `SetUpdatesAttributeName()` along with `_incomingValueBeforeParsing` and other stuff from my pull request, making our custom logic obsolete. Now, the `InputBase` already addresses the issue we were compensating for, simplifying the process to just `CurrentValueAsString = newInputValue;`.

https://github.com/dotnet/aspnetcore/pull/46434

**Comment by hakenr** [comment]
We should be able to drop all these pre-net8 "branches" in November 2024:
https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core

**Comment by hakenr** [comment]
@jirikanda Is this resolved?

**Comment by jirikanda** [comment]
Yes. It was resolved in HxInputDate and HxInputDateRange and I did not find any other *internal parts of components* deriving from HxInputBase<> so I hope it is completely resolved.
